src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/PartitionedRateLimiter.cs

```diff
@@ -0,0 +1,243 @@
+                    limiter = new Lazy<RateLimiter>(() => partition.Factory(partition.PartitionKey));
+                    _limiters.Add(partition.PartitionKey, limiter);
+                    // Cache is invalid now
+                    _cacheInvalid = true;
+                }
+            }
+            return limiter.Value;
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (!disposing)
+            {
+                return;
+            }
+
+            Dictionary<TKey, Lazy<RateLimiter>> limiters;
+            lock (Lock)
+            {
+                if (CommonDisposeUnsynchronized())
+                {
+                    return;
+                }
+
+                limiters = _limiters;
+                // Don't call Clear() as that would clear the local limiters reference as well, which is what we'll be using to call DisposeAsync on all the limiters
+                _limiters = new Dictionary<TKey, Lazy<RateLimiter>>();
+            }
+
+            _timerTask.GetAwaiter().GetResult();
+
+            foreach (KeyValuePair<TKey, Lazy<RateLimiter>> limiter in limiters)
+            {
+                limiter.Value.Value.Dispose();
+            }
+        }
+
+        protected override async ValueTask DisposeAsyncCore()
+        {
+            Dictionary<TKey, Lazy<RateLimiter>> limiters;
+            lock (Lock)
+            {
+                if (CommonDisposeUnsynchronized())
+                {
+                    return;
+                }
+
+                limiters = _limiters;
+                // Don't call Clear() as that would clear the local limiters reference as well, which is what we'll be using to call DisposeAsync on all the limiters
+                _limiters = new Dictionary<TKey, Lazy<RateLimiter>>();
+            }
+
+            await _timerTask.ConfigureAwait(false);
+
+            foreach (KeyValuePair<TKey, Lazy<RateLimiter>> limiter in limiters)
+            {
+                await limiter.Value.Value.DisposeAsync().ConfigureAwait(false);
+            }
+
+            return;
+        }
+
+        // This handles the common state changes that Dipose and DisposeAsync need to do, the individual limiters still need to be Disposed after this call
+        private bool CommonDisposeUnsynchronized()
+        {
+            if (_disposed)
+            {
+                return true;
+            }
+            _disposed = true;
+            _timer.Stop();
+
+            return false;
+        }
+
+        private void ThrowIfDisposed()
+        {
+            if (_disposed)
+            {
+                throw new ObjectDisposedException(nameof(PartitionedRateLimiter));
+            }
+        }
+
+        private static void Replenish(DefaultPartitionedRateLimiter<TResource, TKey> limiter)
+        {
+            List<Lazy<RateLimiter>>? cachedLimiters = Interlocked.Exchange(ref limiter._cachedLimiters, null);
+            if (cachedLimiters is null)
```

Comment:
We can remove the interlocked shenanigans now that there are no overlapping calls, right?